utils/bag_trajectory.py

```diff
@@ -0,0 +1,45 @@
+from abc import abstractmethod
+from collections import defaultdict
+from typing import Collection
+
+import numpy as np
+
+from .registry import MetaParent
+
+
+class BaseBagTrajectoryMemory(metaclass=MetaParent):
+    @abstractmethod
+    def add_to_trajectory(self, bag_ids: Collection, nodes: Collection, infos: Collection):
+        raise NotImplementedError
+
+    @abstractmethod
+    def add_reward_to_trajectory(self, bag_id, reward):
+        raise NotImplementedError
+
+    @abstractmethod
+    def sample_trajectories_for_node(self, node, count):
+        raise NotImplementedError
+
+
+class SharedBagTrajectoryMemory(BaseBagTrajectoryMemory, config_name='shared_path_memory'):
+    _trajectory_by_bag_id = defaultdict(list)
+    _trajectory_by_node = defaultdict(dict)
```

Comment:
Here I think it would be better to name `_start_trajectories_ids_by_node` since we may have multiple for every node. I mean as far as I understand here it should be a mapping from `node_id`, `bag_id` to starting index/indices describing when bag with `bag_id` was in the node with `node_id`. Also, probably in future, there may be cycles so please add TODO near this variable to implement multiple indices support in the future.